radix-engine/src/kernel/kernel.rs

```diff
@@ -406,104 +359,76 @@ where
         } else {
             None
         };
+        let caller = self.current_frame.actor.clone();
 
-        // Filter
-        self.execute_in_mode(ExecutionMode::AuthModule, |system_api| {
-            AuthModule::on_before_frame_start(&actor, system_api)
-        })?;
-
-        // New Call Frame pre-processing
+        // Before push call frame
         {
-            // TODO: Abstract these away
-            self.execute_in_mode(ExecutionMode::TransactionModule, |system_api| {
-                TransactionHashModule::on_call_frame_enter(
-                    &mut call_frame_update,
-                    &actor,
-                    system_api,
-                )
-            })?;
-            self.execute_in_mode(ExecutionMode::LoggerModule, |system_api| {
-                LoggerModule::on_call_frame_enter(&mut call_frame_update, &actor, system_api)
-            })?;
-            self.execute_in_mode(ExecutionMode::AuthModule, |system_api| {
-                AuthModule::on_call_frame_enter(&mut call_frame_update, &actor, system_api)
-            })?;
-            self.execute_in_mode(ExecutionMode::NodeMoveModule, |system_api| {
-                NodeMoveModule::on_call_frame_enter(
-                    &mut call_frame_update,
-                    &actor.identifier,
-                    system_api,
-                )
+            self.execute_in_mode(ExecutionMode::KernelModule, |api| {
+                KernelModuleMixer::before_push_frame(api, &actor, &mut call_frame_update)
             })?;
-
-            self.module
-                .pre_execute_invocation(
-                    &actor,
-                    &call_frame_update,
-                    &mut self.current_frame,
-                    &mut self.heap,
-                    &mut self.track,
-                )
-                .map_err(RuntimeError::ModuleError)?;
-            self.id_allocator.pre_execute_invocation();
         }
 
-        // Call Frame Push
+        // Push call frame
         {
+            self.id_allocator.push();
+
             let frame = CallFrame::new_child_from_parent(
                 &mut self.current_frame,
-                actor,
+                actor.clone(),
                 call_frame_update,
             )?;
             let parent = mem::replace(&mut self.current_frame, frame);
             self.prev_frame_stack.push(parent);
         }
 
         // Execute
-        let (output, update) = self.execute_in_mode(ExecutionMode::Application, |system_api| {
-            executor.execute(system_api)
-        })?;
+        let (output, update) = {
+            // Handle execution start
+            self.execute_in_mode(ExecutionMode::KernelModule, |api| {
+                KernelModuleMixer::on_execution_start(api, &caller)
+            })?;
+
+            // Run
+            let (output, mut update) =
+                self.execute_in_mode(ExecutionMode::Client, |api| executor.execute(api))?;
+
```

Comment:
The auto-drop locks should be moved to here. Otherwise, there's two drop locks in a row below.

Reply:
Nice catch. I will clean it up!